dev/test/index.ts

```diff
@@ -1273,3 +1296,389 @@ describe('getAll() method', () => {
+          return streamPromise;
+        },
+        batchWrite: request => {
+          const expected = createRequest([
+            deleteOp('a'),
+            deleteOp('b'),
+            deleteOp('c'),
+            deleteOp('d'),
+          ]);
+          try {
+            expect(request.writes).to.deep.equal(expected.writes);
+          } catch (e) {
+            batchWriteError = e;
+          }
+          return response(
+            mergeResponses(expected.writes!.map(() => successResponse(1)))
+          );
+        },
+      };
+      firestore = await createInstance(overrides);
+      await firestore.recursiveDelete(firestore.collection('letters'));
+      expect(batchWriteError, 'batchWrite should not have errored').to.be
+        .undefined;
+    });
+
+    it('handles multiple calls to recursiveDelete()', async () => {
+      let requestCounter = 0;
+      const docIds = ['a', 'b', 'c'];
+      const streamItems = docIds.map(docId => [result(docId)]);
+      const expected = docIds.map(docId => createRequest([deleteOp(docId)]));
+      const responses = docIds.map(() => successResponse(1));
+
+      const overrides: ApiOverride = {
+        runQuery: () => {
+          return stream(...streamItems[requestCounter]);
+        },
+        batchWrite: request => {
+          try {
+            expect(request.writes).to.deep.equal(
+              expected[requestCounter]!.writes
+            );
+          } catch (e) {
+            batchWriteError = e;
+          }
+          const responsePromise = response(responses[requestCounter]);
+          requestCounter++;
+          return responsePromise;
+        },
+      };
+      firestore = await createInstance(overrides);
+      await firestore.recursiveDelete(firestore.collection('a'));
+      await firestore.recursiveDelete(firestore.collection('b'));
+      await firestore.recursiveDelete(firestore.collection('c'));
+      expect(batchWriteError, 'batchWrite should not have errored').to.be
+        .undefined;
+    });
+
+    it('accepts references with converters', async () => {
+      const overrides: ApiOverride = {
+        runQuery: () => {
+          return stream();
+        },
+        // Include response for deleting the provided document reference.
+        batchWrite: () => {
+          return response(successResponse(1));
+        },
+      };
+      firestore = await createInstance(overrides);
+      await firestore.recursiveDelete(
+        firestore.doc('root/doc').withConverter(postConverter)
+      );
+      await firestore.recursiveDelete(
+        firestore.collection('root').withConverter(postConverter)
+      );
+    });
+  });
+
+  describe('BulkWriter instance', () => {
+    it('uses custom BulkWriter instance if provided', async () => {
+      firestore = await instantiateInstance([['a', 'b', 'c']]);
+      let callbackCount = 0;
+      const bulkWriter = firestore.bulkWriter();
+      bulkWriter.onWriteResult(() => {
+        callbackCount++;
```

Comment:
Do we want to add a test that verifies that we return the expected list of DocumentReferences?